client/src/lib/paymentCalc.ts

```diff
@@ -1,56 +1,115 @@
+
   // Get proofVerificationGas from contract
-  let proofVerificationGas;
-  try {
-    proofVerificationGas = await publicClient.readContract({
-      address: axiomV2QueryAddr,
-      abi: getAxiomV2Abi(AbiType.Query),
-      functionName: "proofVerificationGas",
-      args: [],
-    }) as bigint; // in gas units
-  } catch (e) {
-    console.log(`Unable to read proofVerificationGas from contract`);
-  }
-  proofVerificationGas = BigInt(proofVerificationGas ?? 0);
-  if (proofVerificationGas === 0n) {
-    proofVerificationGas = ClientConstants.FALLBACK_PROOF_VERIFICATION_GAS;
-  }
+  const proofVerificationGas = await readContractValue(
+    publicClient,
+    axiomV2QueryAddr,
+    getAxiomV2Abi(AbiType.Query),
+    "proofVerificationGas",
+    [],
+    ClientConstants.FALLBACK_PROOF_VERIFICATION_GAS
+  );
 
   // Get axiomQueryFee from contract
-  let axiomQueryFee;
-  try {
-    axiomQueryFee = await publicClient.readContract({
-      address: axiomV2QueryAddr,
-      abi: getAxiomV2Abi(AbiType.Query),
-      functionName: "axiomQueryFee",
-      args: [],
-    }) as bigint; // in wei
-  } catch (e) {
-    console.log(`Unable to read axiomQueryFee from contract`);
-  }
-  axiomQueryFee = BigInt(axiomQueryFee ?? 0);
-  if (axiomQueryFee === 0n) {
-    axiomQueryFee = ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI;
-  }
+  let axiomQueryFee = await readContractValue(
+    publicClient,
+    axiomV2QueryAddr,
+    getAxiomV2Abi(AbiType.Query),
+    "axiomQueryFee",
+    [],
+    ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI
+  );
 
-  // Convert callback gas limit to wei
-  const callbackGasLimit = BigInt(options?.callbackGasLimit ?? ClientConstants.DEFAULT_CALLBACK_GAS_LIMIT);
+  if (isOpStackChain(chainId) || isArbitrumChain(chainId) || isScrollChain(chainId)) {
+    // Get the projected callback cost
+    const projectedCallbackCost = await getProjectedL2CallbackCost(chainId, publicClient, maxFeePerGas, callbackGasLimit, proofVerificationGas);
+
+    // Get overrideAxiomQueryFee from either equation or options
+    let overrideAxiomQueryFee: bigint;
+    if (options?.overrideAxiomQueryFee !== undefined) {
+      overrideAxiomQueryFee = BigInt(options.overrideAxiomQueryFee);
+    } else {
+      // overrideAxiomQueryFee = AXIOM_QUERY_FEE + projectedCallbackCost - maxFeePerGas * (callbackGasLimit + proofVerificationGas)
+      overrideAxiomQueryFee = ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI + 
+        projectedCallbackCost - maxFeePerGas * (callbackGasLimit + proofVerificationGas);
+    }
+
+    // max(overrideAxiomQueryFee, axiomQueryFee)
+    if (overrideAxiomQueryFee > axiomQueryFee) {
+      axiomQueryFee = overrideAxiomQueryFee;
+    }
+
+    // Calculate payment
+    let payment = axiomQueryFee + maxFeePerGas * (proofVerificationGas + callbackGasLimit);
+    const minimumPayment = projectedCallbackCost + ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI;
+    if (payment < minimumPayment) {
```

Comment:
I think we'll want to error in this case, since it means something was misconfigured